File: 13-Graph-Algorithms/10-Network-Flow/java/MaxFlow.java
import java.util.*;
public class MaxFlow {
    static int[][] cap; static int[] parent;
    static boolean bfs(int s,int t,int n){parent=new int[n];Arrays.fill(parent,-1);parent[s]=-2;Queue<Integer> q=new LinkedList<>();q.add(s);while(!q.isEmpty()){int u=q.poll();for(int v=0;v<n;v++)if(parent[v]==-1&&cap[u][v]>0){parent[v]=u;if(v==t)return true;q.add(v);}}return false;}
    static int maxFlow(int s,int t,int n){int flow=0;while(bfs(s,t,n)){int pf=Integer.MAX_VALUE,u=t;while(u!=s){pf=Math.min(pf,cap[parent[u]][u]);u=parent[u];}u=t;while(u!=s){cap[parent[u]][u]-=pf;cap[u][parent[u]]+=pf;u=parent[u];}flow+=pf;}return flow;}
    public static void main(String[] args){cap=new int[6][6];cap[0][1]=16;cap[0][2]=13;cap[1][3]=12;cap[2][4]=14;cap[3][5]=20;cap[4][5]=4;System.out.println("Max Flow: "+maxFlow(0,5,6));}
}

File: 13-Graph-Algorithms/10-Network-Flow/cpp/max_flow.cpp
/* Max Flow (Edmonds-Karp / BFS-based Ford-Fulkerson) | Time: O(VE²) */
#include <bits/stdc++.h>
using namespace std;
const int MAXN=105, INF=1e9;
int cap[MAXN][MAXN], parent_[MAXN];
bool bfs(int s,int t,int n){
    memset(parent_,-1,sizeof(parent_)); parent_[s]=-2;
    queue<pair<int,int>> q; q.push({s,INF});
    while(!q.empty()){auto[u,flow]=q.front();q.pop();
        for(int v=0;v<n;v++){if(parent_[v]==-1&&cap[u][v]>0){parent_[v]=u;int nf=min(flow,cap[u][v]);if(v==t)return true;q.push({v,nf});}}}
    return false;
}
int maxFlow(int s,int t,int n){
    int flow=0;
    while(bfs(s,t,n)){
        int pathFlow=INF,u=t;
        while(u!=s){int p=parent_[u];pathFlow=min(pathFlow,cap[p][u]);u=p;}
        u=t; while(u!=s){int p=parent_[u];cap[p][u]-=pathFlow;cap[u][p]+=pathFlow;u=p;}
        flow+=pathFlow;
    }
    return flow;
}
int main(){
    cap[0][1]=16;cap[0][2]=13;cap[1][2]=10;cap[1][3]=12;cap[2][1]=4;cap[2][4]=14;cap[3][2]=9;cap[3][5]=20;cap[4][3]=7;cap[4][5]=4;
    cout<<"Max Flow: "<<maxFlow(0,5,6)<<endl; // 23
}
